runtime/drivers/blob/blobdownloader.go

```diff
@@ -148,138 +169,86 @@ func (it *blobIterator) Close() error {
+		go func() {
+			it.init()
+		}()
+	}
 	if !it.opts.KeepFilesUntilClose {
 		// delete files created in last iteration
-		fileutil.ForceRemoveFiles(it.localFiles)
+		fileutil.ForceRemoveFiles(it.lastBatch)
 	}
 
-	// new slice creation is not necessary on every iteration
-	// but there may be cases where n in first batch is different from n in next batch
-	// to keep things easy creating a new slice every time
-	it.localFiles = make([]string, 0)
-	g, grpCtx := errgroup.WithContext(it.ctx)
-	g.SetLimit(_concurrentBlobDownloadLimit)
-
+	it.lastBatch = it.lastBatch[:0]
 	var totalSizeInBytes int64
-	for ; it.index < len(it.objects) && totalSizeInBytes < sizeInBytes; it.index++ {
-		obj := it.objects[it.index]
-		totalSizeInBytes += obj.obj.Size
-		// need to create file by maintaining same dir path as in glob for hivepartition support
-		filename := filepath.Join(it.tempDir, obj.obj.Key)
-		it.localFiles = append(it.localFiles, filename)
-		g.Go(func() error {
-			if err := os.MkdirAll(filepath.Dir(filename), os.ModePerm); err != nil {
-				return err
-			}
-
-			file, err := os.OpenFile(filename, os.O_RDWR|os.O_CREATE, os.ModePerm)
-			if err != nil {
-				return err
-			}
-			defer file.Close()
-
-			ext := filepath.Ext(obj.obj.Key)
-			partialReader, isPartialDownloadSupported := _partialDownloadReaders[ext]
-			downloadFull := obj.full || !isPartialDownloadSupported
-
-			// Collect metrics of download size and time
-			startTime := time.Now()
-			defer func() {
-				size := obj.obj.Size
-				st, err := file.Stat()
-				if err == nil {
-					size = st.Size()
-				}
-
-				duration := time.Since(startTime)
-				it.logger.Info("download complete", zap.String("object", obj.obj.Key), zap.Duration("duration", duration), observability.ZapCtx(it.ctx))
-				drivers.RecordDownloadMetrics(grpCtx, &drivers.DownloadMetrics{
-					Connector: "blob",
-					Ext:       ext,
-					Partial:   !downloadFull,
-					Duration:  duration,
-					Size:      size,
-				})
-			}()
-
-			// download full file
-			if downloadFull {
-				return downloadObject(grpCtx, it.bucket, obj.obj.Key, file)
-			}
-			// download partial file
-			// check if, for smaller size we can download entire file
-			switch partialReader {
-			case "parquet":
-				return downloadParquet(grpCtx, it.bucket, obj.obj, obj.extractOption, file)
-			case "csv":
-				return downloadText(grpCtx, it.bucket, obj.obj, &textExtractOption{extractOption: obj.extractOption, hasCSVHeader: true}, file)
-			case "json":
-				return downloadText(grpCtx, it.bucket, obj.obj, &textExtractOption{extractOption: obj.extractOption, hasCSVHeader: false}, file)
-			default:
-				// should not reach here
-				panic(fmt.Errorf("partial download not supported for extension %q", ext))
+	var err error
+	for totalSizeInBytes < it.opts.BatchSizeBytes {
+		fileName, ok := <-it.downloadedFiles
+		if !ok { // channel is closed
+			err = it.err
+			if errors.Is(err, it.err) {
+				err = nil
```

Comment:
I don't understand this

Reply:
We don't want to return `io.EOF` from `Next`. Clients check `hasNext` to check if more elements are present and any error returned from next is a genuine error.

Reply:
I still don't understand – the code reads like:
```go
a := 10
if a == 10 {
  a = 0
}
```